File: 658-find-k-closest-elements/658-find-k-closest-elements.scala
import scala.collection.mutable._

object Solution {
    def findClosestElements(arr: Array[Int], k: Int, x: Int): List[Int] = {
        if (k >= arr.length) arr.toList
        else {
            val index = binSearch(arr, x)
            
            var i = index
            var j = index + 1

            for(l <- 1 to k) {
                if (i < 0) j += 1
                else if (j == arr.length) i -= 1
                else if (isCloserThan(x, arr(i), arr(j))) i -= 1
                else j += 1
            }
            
            arr.slice(i + 1, j).toList
        }
    }

    private def isCloserThan(target: Int, i: Int, j: Int): Boolean = {
        val di = Math.abs(i - target)
        val dj = Math.abs(j - target)
        if (di < dj || (di == dj && i <= j)) true else false
    }

    private def binSearch(arr: Array[Int], x: Int): Int = {
        var start = 0
        var end = arr.length - 1

        while(start <= end) {
            val mid = (start + end) / 2
            val v = arr(mid)

            if (x < v) end = mid - 1
            else if (x >= v) start = mid + 1
        }

        end
    }
}

File: 658-find-k-closest-elements/658-find-k-closest-elements.java
class Solution {
    public List<Integer> findClosestElements(int[] arr, int k, int x) {
        int idx = binSearch(arr, x);
        
        int r = idx;
        int l = idx - 1;
        
        List<Integer> res = new ArrayList<>();
        
        for(int i = 0; i < k; i++) {
            if (r >= arr.length) {
                l--;
                continue;
            } else if (l < 0) {
                r++;
                continue;
            }
            
            if (Math.abs(arr[r] - x) < Math.abs(x - arr[l])) r++;
            else l--;
        }
        
        for(int i = 0; i < k; i++) {
            res.add(arr[i + l + 1]);
        }
        
        return res;
    }
    
    int binSearch(int[] arr, int x) {
        int s = 0;
        int e = arr.length - 1;
        
        while(s <= e) {
            int m = s + (e - s) / 2;
            if (arr[m] < x) s = m + 1;
            else e = m - 1;
        }
        
        return s;
    }
}